File: 2018/p14.rb
require './helper'

class Recipe
  def initialize
    @pos1, @prev1 = 0, 0
    @pos2, @prev2 = 1, 1
    # @r_scores = [3,7]
    @r_scores = "37"
    @num_ticks = 0
  end

  attr_reader :pos1, :pos2, :r_scores, :r_scoresstr,:prev1, :prev2, :num_ticks

  def tick
    new_recipe = (r_scores[pos1].to_i + r_scores[pos2].to_i).to_s
    @r_scores += new_recipe
    @prev1, @prev2 = pos1, pos2
    @pos1, @pos2 = (pos1 + 1 + r_scores[pos1].to_i)%r_scores.size, (pos2 + 1 + r_scores[pos2].to_i)%r_scores.size
    @num_ticks += 1
    r_scores
  end

  def predict10(n)
    n.times{ tick }
    10.times{ tick }
    score = r_scores[n...n+10]
    puts score 
    score
  end

  def part2(num)
    str = num.to_s
    num_rec_pre = 0
    len = str.size
    loop do 
      tick  
      if r_scores[-len..-1] == str
        num_rec_pre = r_scores.size - len
        break
      elsif r_scores[-len-1..-2] == str
        num_rec_pre = r_scores.size - len -1
        break
      end
    end

    puts num_rec_pre
    num_rec_pre
  end

end


### these approaches are pretty straight forward. doesn't work for part 2 though. Need to figure out the pattern. input 765071.

File: 2018/p14.java

public class Aoc14 {
  int pos1 = 0; 
  int pos2 = 1; 
  String r_scores = "37";
  
  public static void main(String[] args) {
    
  }
  
  
  public String predict10(int n) {
    for(int i = 1; i<= n + 10; i++) {
      tick();
      
    }
    String score = r_scores.substring(n, n+10);
    System.out.println(score);
    return score;
  }
  
  public void tick() {
    String new_recipe = "" + (Character.getNumericValue(r_scores.charAt(pos1)) + Character.getNumericValue(r_scores.charAt(pos2)));
    r_scores += new_recipe;
    pos1 = (pos1 + 1 + Character.getNumericValue(r_scores.charAt(pos1)))%r_scores.length() ;
    pos2 = (pos2 + 1 + Character.getNumericValue(r_scores.charAt(pos2)))%r_scores.length() ;
  }
  

  // This doesn't work. It probably requires a few years to get the answer with input 765071. 
  // I'm sure there's some pattern somwhere that prevents unnecessary computation, I just have to figure thaat out. 
  public int part2(int num) {
    String str = "" + num;
    int num_rec_pre = 0;
      int len = str.length();
      while(true) {
        tick();
        int l = r_scores.length();
        if(r_scores.substring(Math.floorMod((l-len), l),l).equals(str)) {
          num_rec_pre = r_scores.length() - len;
          break;
        }
        else if(r_scores.substring(Math.floorMod((l-len-1),l),l-1).equals(str)) {
          num_rec_pre = r_scores.length() - len -1;
          break;
        }
      }
      System.out.println(num_rec_pre);
      return num_rec_pre;
  }
  
}
